contract/models/contract_line.py

```diff
@@ -512,31 +512,57 @@ def _prepare_invoice_line(self, invoice_id=False):
         )
         return invoice_line_vals
 
-    @api.multi
-    def _get_invoiced_period(self):
-        self.ensure_one()
+    @api.model
+    def _get_next_invoiced_period(
+        self,
+        date_start,
+        date_end,
+        last_date_invoiced,
+        recurring_next_date,
+        recurring_interval,
+        recurring_rule_type,
+        recurring_invoicing_type,
+    ):
         first_date_invoiced = (
-            self.last_date_invoiced + relativedelta(days=1)
-            if self.last_date_invoiced
-            else self.date_start
+            last_date_invoiced + relativedelta(days=1)
+            if last_date_invoiced
+            else date_start
         )
-        if self.recurring_rule_type == 'monthlylastday':
-            last_date_invoiced = self.recurring_next_date
+        if recurring_rule_type == 'monthlylastday':
+            last_date_invoiced = recurring_next_date
         else:
-            if self.recurring_invoicing_type == 'pre-paid':
+            if recurring_invoicing_type == 'pre-paid':
                 last_date_invoiced = (
-                    self.recurring_next_date
+                    recurring_next_date
                     + self.get_relative_delta(
-                        self.recurring_rule_type, self.recurring_interval
+                        recurring_rule_type, recurring_interval
                     )
                     - relativedelta(days=1)
                 )
             else:
-                last_date_invoiced = self.recurring_next_date - relativedelta(
+                last_date_invoiced = recurring_next_date - relativedelta(
                     days=1
                 )
-        if self.date_end and self.date_end < last_date_invoiced:
-            last_date_invoiced = self.date_end
+        if date_end and date_end < last_date_invoiced:
+            last_date_invoiced = date_end
+        recurring_next_date = recurring_next_date + self.get_relative_delta(
+            recurring_rule_type, recurring_interval
+        )
+        return first_date_invoiced, last_date_invoiced, recurring_next_date
+
+    @api.multi
+    def _get_invoiced_period(self):
+        self.ensure_one()
+        dates = self._get_next_invoiced_period(
+            self.date_start,
+            self.date_end,
+            self.last_date_invoiced,
+            self.recurring_next_date,
+            self.recurring_interval,
+            self.recurring_rule_type,
+            self.recurring_invoicing_type,
```

Comment:
@sbejaoui for the purpose of reusing in the forecasting algorihtm it would be slightly easier to read if _get_next_invoiced_period remains a multi/ensure_one method, taking only last_date_invoiced and recurring_next_date as arguments.

Reply:
yes